### PR TITLE
WIP: New version

### PR DESCRIPTION
**Pull request by jetexe**
| Q             | A
| ------------- | ---
| Bug fix?      | Yes
| New feature?  | Yes

## Description

New version

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I wrote unit tests for my code _(if tests is required for my changes)_
- [ ] I have made changes in [CHANGELOG.md](https://github.com/avto-dev/cloud-payments-laravel/blob/master/CHANGELOG.md) file

### CODE REVIEW

**Comment by tarampampam** [comment]
```yaml
preset: laravel

enabled:
  - strict
  - align_double_arrow
  - dir_constant
  - linebreak_after_opening_tag
  - mb_str_functions
  - no_short_echo_tag
  - non_printable_character
  - ordered_class_elements
  - phpdoc_add_missing_param_annotation
  - phpdoc_align
  - phpdoc_annotation_without_dot
  - phpdoc_order
  - phpdoc_separation
  - phpdoc_return_self_reference
  - align_equals
  - concat_with_spaces
  - length_ordered_imports

disabled:
  - unalign_equals
  - concat_without_spaces
  - declare_equal_normalize
  - alpha_ordered_imports
```

File: src/Receipts/Receipt.php [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+    {
+        return $this->items;
+    }
+
+    /**
+     * @param Item[] $items
+     *
+     * @return $this
+     */
+    public function setItems(array $items): self
```
Comment (tarampampam) [inline src/Receipts/Receipt.php]: assert that $item is instanceof Item

File: src/Receipts/Receipt.php [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+            'email'            => $this->email,
+            'phone'            => $this->phone,
+            'isBso'            => $this->is_bso,
+            'amounts'          => [
+                'electronic'     => \number_format((float) $this->electronic_amount, 2, '.', ''),
+                'advancePayment' => \number_format((float) $this->advance_payment_amount, 2, '.', ''),
+                'credit'         => \number_format((float) $this->credit_amount, 2, '.', ''),
+                'provision'      => \number_format((float) $this->provision_amount, 2, '.', ''),
+            ],
+        ], static function ($value) {
```
Comment (tarampampam) [inline src/Receipts/Receipt.php]: as idea: `\\is_null`

File: src/References/Currency.php [hunk 3]
```diff
@@ -0,0 +1,111 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace AvtoDev\CloudPayments\References;
+
+/**
+ * @see https://developers.cloudpayments.ru/#spisok-valyut
+ */
+class Currency
```
Comment (tarampampam) [inline src/References/Currency.php]: interface?

File: src/Requests/AbstractRequestBuilder.php [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        $request = new Request('POST', $uri);
+
+        if ($this->request_id !== '') {
+            $request = $request->withHeader('X-Request-ID', $this->request_id);
+        }
+
+        /** @var RequestInterface $request */
+        $request = $request->withHeader('Content-Type', 'application/json');
+
+        $request_data = \array_filter($this->getRequestParams(), static function ($value) {
```
Comment (tarampampam) [inline src/Requests/AbstractRequestBuilder.php]: move to `$this->filterPayload(array $payload)`

File: src/ServiceProvider.php [hunk 5]
```diff
@@ -4,13 +4,12 @@
...
-use AvtoDev\CloudPayments\Client\Client;
-use AvtoDev\CloudPayments\Client\ClientInterface;
-use GuzzleHttp\ClientInterface as GuzzleHttpClientInterface;
+use GuzzleHttp\Client as GuzzleClient;
 use Illuminate\Contracts\Config\Repository;
 use Illuminate\Contracts\Container\Container;
+use Illuminate\Support\ServiceProvider as BaseServiceProvider;
 
-class ServiceProvider extends \Illuminate\Support\ServiceProvider
+class ServiceProvider extends BaseServiceProvider
```
Comment (tarampampam) [inline src/ServiceProvider.php]: rm? :1st_place_medal:

Comment (tarampampam) [reply src/ServiceProvider.php]: :smoking:

File: src/Requests/Traits/PaymentRequestTrait.php [hunk 6]
```diff
@@ -0,0 +1,271 @@
+<?php
+
```
Comment (tarampampam) [inline src/Requests/Traits/PaymentRequestTrait.php]: `declare(strict_types = 1);` in all places (includes tests)

File: tests/Feature/PaymentsCardsChargeTest.php [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+use AvtoDev\CloudPayments\Exceptions\InvalidRequestException;
+use AvtoDev\CloudPayments\Requests\Payments\Cards\CardsAuthRequestBuilder;
+use AvtoDev\CloudPayments\ResponseParser;
+
+/**
+ * @coversNothing
+ */
+class PaymentsCardsChargeTest extends AbstractFeatureTestCase
+{
+    public function testInvelidFormatTest()
```
Comment (tarampampam) [inline tests/Feature/PaymentsCardsChargeTest.php]: `invalidFormatTest(): void`

File: tests/Feature/TestRequestTest.php [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+     */
+    public function testIdempotenceTest(): void
+    {
+        $request = new TestRequestBuilder;
+
+        $request->setRequestId('some');
+
+        $response = $this->cloud_payments_client->send($request->buildRequest());
+
+        $message = ResponseParser::parseSimpleResponse($response)->getMessage();
```
Comment (tarampampam) [inline tests/Feature/TestRequestTest.php]: Remove Response parser

File: tests/Feature/TestRequestTest.php [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+use AvtoDev\CloudPayments\ResponseParser;
+
+/**
+ * @coversNothing
+ */
+class TestRequestTest extends AbstractFeatureTestCase
+{
+    /**
+     * @throws \GuzzleHttp\Exception\GuzzleException
+     * @throws \Tarampampam\Wrappers\Exceptions\JsonEncodeDecodeException
```
Comment (tarampampam) [inline tests/Feature/TestRequestTest.php]: remove throws all places

File: tests/Unit/AbstractUnitTestCase.php [hunk 10]
```diff
@@ -0,0 +1,11 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace AvtoDev\Tests\Unit;
+
+use AvtoDev\Tests\AbstractTestCase;
+
+abstract class AbstractUnitTestCase extends AbstractTestCase
```
Comment (tarampampam) [inline tests/Unit/AbstractUnitTestCase.php]: Remove

File: tests/Unit/Receipt/ReceiptTest.php [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+        $item = new Item;
+
+        $item->setLabel(Str::random());
+
+        $items = [
+            $item,
+            new \stdClass,
+        ];
+
+        $this->receipt->setItems($items);
```
Comment (tarampampam) [inline tests/Unit/Receipt/ReceiptTest.php]: remove `setItemsMethod`

File: tests/Unit/Requests/AbstractRequestBuilderTest.php [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+declare(strict_types = 1);
+
+namespace AvtoDev\Tests\Unit\Requests;
+
+use AvtoDev\CloudPayments\Requests\AbstractRequestBuilder;
+use AvtoDev\Tests\Unit\AbstractUnitTestCase;
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\UriInterface;
+
+abstract class AbstractRequestBuilderTest extends AbstractUnitTestCase
```
Comment (tarampampam) [inline tests/Unit/Requests/AbstractRequestBuilderTest.php]: AbstractRequestBuilderTestCaes

File: .github/workflows/execute-tests.yml [hunk 13]
```diff
@@ -16,10 +16,10 @@ jobs: # Docs: <https://help.github.com/en/articles/workflow-syntax-for-github-ac
     strategy:
       matrix:
         setup: ['basic', 'lowest']
-        php: ['7.1', '7.2', '7.3']
+        php: ['7.1', '7.2', '7.3','7.4']
```
[the remark this answers is not in the document]
Comment (jetexe) [reply .github/workflows/execute-tests.yml]: Fixed

File: Dockerfile [hunk 14]
```diff
@@ -1,6 +1,11 @@
 FROM composer:1.8.6 AS composer
 
-FROM php:7.1.0-alpine
+FROM php:7.1-alpine
```
[the remark this answers is not in the document]
Comment (jetexe) [reply Dockerfile]: Fixed

File: Makefile [hunk 15]
```diff
@@ -2,12 +2,14 @@
 # Makefile readme (ru): <http://linux.yaroslavl.ru/docs/prog/gnu_make_3-79_russian_manual.html>
 # Makefile readme (en): <https://www.gnu.org/software/make/manual/html_node/index.html#SEC_Contents>
 
-dc_bin := $(shell command -v docker-compose 2> /dev/null)
+docker_bin := $(shell command -v docker 2> /dev/null)
```
[the remark this answers is not in the document]
Comment (jetexe) [reply Makefile]: Fixed

File: composer.json [hunk 16]
```diff
@@ -1,24 +1,29 @@
 {
     "name": "avto-dev/cloud-payments-laravel",
     "description": "Cloud Payments PHP-client",
-    "keywords": ["cloud-payments"],
+    "keywords": [
+        "cloud-payments"
```
[the remark this answers is not in the document]
Comment (jetexe) [reply composer.json]: Fixed

File: composer.json [hunk 17]
```diff
@@ -1,24 +1,29 @@
...
+    ],
     "type": "library",
     "license": "MIT",
     "require": {
         "php": ">=7.1",
```
[the remark this answers is not in the document]
Comment (jetexe) [reply composer.json]: Fixed

File: composer.json [hunk 18]
```diff
@@ -1,24 +1,29 @@
...
         "psr/http-client": "~1.0",
         "tarampampam/wrappers-php": "^1.2"
     },
     "require-dev": {
-        "phpunit/phpunit": "^6.4 || ~7.5",
-        "phpstan/phpstan": "^0.11.19",
+        "fzaninotto/faker": "^1.9",
         "laravel/laravel": "^5.5 || ^6.0",
-        "fzaninotto/faker": "^1.9"
+        "phpstan/phpstan": "^0.11.19",
```
[the remark this answers is not in the document]
Comment (jetexe) [reply composer.json]: Fixed

File: src/Client.php [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+     */
+    protected $client;
+
+    /**
+     * @var Config
+     */
+    protected $config;
+
+    /**
+     * Client constructor.
```
[the remark this answers is not in the document]
Comment (jetexe) [reply src/Client.php]: Fixed

File: src/Client.php [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+     *
+     * @throws GuzzleException
+     *
+     * @return ResponseInterface
+     */
+    public function send(RequestInterface $request): ResponseInterface
+    {
+        try {
+            return $this->client->send($request, $this->getRequestOptions());
+        } catch (\Exception $exception) {
```
[the remark this answers is not in the document]
Comment (jetexe) [reply src/Client.php]: Fixed

File: src/Client.php [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+            throw CloudPaymentsRequestException::wrapException($request, $exception);
+        }
+    }
+
+    protected function getRequestOptions(): array
+    {
+        return [
+            'auth' => [$this->config->getPublicId(), $this->config->getApiKey()],
```
[the remark this answers is not in the document]
Comment (jetexe) [reply src/Client.php]: Fixed

File: src/Config.php [hunk 22]
```diff
@@ -0,0 +1,38 @@
...
+     * @var string
+     */
+    protected $api_key;
+
+    /**
+     * @var string
+     */
+    protected $public_id;
+
+    public function __construct(array $config)
```
[the remark this answers is not in the document]
Comment (jetexe) [reply src/Config.php]: Fixed

File: src/Receipts/Receipt.php [hunk 23]
```diff
@@ -0,0 +1,306 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace AvtoDev\CloudPayments\Receipts;
+
+use AvtoDev\CloudPayments\References\TaxationSystem;
+
+class Receipt
```
[the remark this answers is not in the document]
Comment (jetexe) [reply src/Receipts/Receipt.php]: Fixed

File: src/Receipts/Receipt.php [hunk 24]
```diff
@@ -0,0 +1,306 @@
...
+    {
+        return $this->calculation_place;
+    }
+
+    /**
+     * @param string|null $calculation_place
+     *
+     * @return $this
+     */
+    public function setCalculationPlace(?string $calculation_place): self
```
[the remark this answers is not in the document]
Comment (jetexe) [reply src/Receipts/Receipt.php]: Fixed

File: src/Receipts/Receipt.php (hunk 1 above)
Comment (jetexe) [reply src/Receipts/Receipt.php]: Removed

File: src/Receipts/Receipt.php (hunk 2 above)
Comment (jetexe) [reply src/Receipts/Receipt.php]: nope, we need filter **not** null values

File: src/Requests/AbstractRequestBuilder.php (hunk 4 above)
Comment (jetexe) [reply src/Requests/AbstractRequestBuilder.php]: Fixed

File: tests/Feature/PaymentsCardsChargeTest.php (hunk 7 above)
Comment (jetexe) [reply tests/Feature/PaymentsCardsChargeTest.php]: removed

File: tests/Feature/TestRequestTest.php (hunk 8 above)
Comment (jetexe) [reply tests/Feature/TestRequestTest.php]: Fixed

File: tests/Feature/TestRequestTest.php (hunk 9 above)
Comment (jetexe) [reply tests/Feature/TestRequestTest.php]: Fixed

File: tests/Unit/AbstractUnitTestCase.php (hunk 10 above)
Comment (jetexe) [reply tests/Unit/AbstractUnitTestCase.php]: Fixed

File: tests/Unit/Receipt/ReceiptTest.php (hunk 11 above)
Comment (jetexe) [reply tests/Unit/Receipt/ReceiptTest.php]: Fixed

File: tests/Unit/Requests/AbstractRequestBuilderTest.php (hunk 12 above)
Comment (jetexe) [reply tests/Unit/Requests/AbstractRequestBuilderTest.php]: Fixed

File: src/Requests/Traits/PaymentRequestTrait.php (hunk 6 above)
Comment (jetexe) [reply src/Requests/Traits/PaymentRequestTrait.php]: Fixed

File: composer.json (hunk 16 above)
[the remark this answers is not in the document]
Comment (jetexe) [reply composer.json]: Fixed

File: tests/Feature/AbstractFeatureTestCase.php [hunk 25]
```diff
@@ -0,0 +1,29 @@
...
+    /**
+     * @var array
+     */
+    protected $config = [];
+
+    public function setUp(): void
+    {
+        parent::setUp();
+
+        $this->config = require __DIR__ . '/config.php';
```
[the remark this answers is not in the document]
Comment (jetexe) [reply tests/Feature/AbstractFeatureTestCase.php]: fixed

File: src/References/Currency.php [hunk 26]
```diff
@@ -2,15 +2,15 @@
...
 
 /**
  * @see https://developers.cloudpayments.ru/#spisok-valyut
  */
-class Currency
+interface Currency
 {
     /**
-     * Russian ruble.
+     * Russian Rouble.
```
[the remark this answers is not in the document]
Comment (jetexe) [reply src/References/Currency.php]: Collapsed  everywhere

File: composer.json [hunk 27]
```diff
@@ -1,24 +1,29 @@
...
     "type": "library",
     "license": "MIT",
     "require": {
-        "php": ">=7.1",
+        "php": ">=7.1.3",
         "ext-json": "*",
-        "guzzlehttp/guzzle": "~6.0",
+        "guzzlehttp/guzzle": "~6.4.1",
+        "guzzlehttp/psr7": "^1.4",
         "illuminate/config": "^5.5 || ^6.0",
```
[the remark this answers is not in the document]
Comment (jetexe) [reply composer.json]: removed

File: config/services.php [hunk 28]
```diff
@@ -4,7 +4,7 @@
 
 return [
```
[the remark this answers is not in the document]
Comment (jetexe) [reply config/services.php]: Removed, but `README.md` not actualized

File: Makefile [hunk 29]
```diff
@@ -7,7 +7,7 @@ dc_bin := $(shell command -v docker-compose 2> /dev/null)
 SHELL = /bin/sh
 RUN_APP_ARGS = --rm --user "$(shell id -u):$(shell id -g)" app
 
-.PHONY : help build latest install lowest test test-cover shell clean
+.PHONY : help build latest install lowest test test-cover shell\ clean
```
[the remark this answers is not in the document]
Comment (jetexe) [reply Makefile]: Fixed

File: src/Client.php [hunk 30]
```diff
@@ -0,0 +1,65 @@
...
+    {
+        $request = $this->updateRequest($request);
+        try {
+            return $this->client->send($request);
+        } catch (GuzzleException $exception) {
+            /** @var \Exception $exception */
+            throw CloudPaymentsRequestException::wrapException($request, $exception);
+        }
+    }
+
```
[the remark this answers is not in the document]
Comment (jetexe) [reply src/Client.php]: fixed

**Pull request merged**